proposals/0003-external-calls.md

```diff
@@ -0,0 +1,344 @@
+
+## Motivation
+Calls to untrusted contracts can introduce several unexpected risks and errors. External calls may execute malicious code in that contract _or any other contract_ that it depends upon. As such, **every** external call should be treated as a security risk.
+
+However external calls are necessary to accomplish key features of smart contracts, including:
+- Paying other users
+- Interacting with other Contracts e.g. Tokens or Wallets
+
+There have been several published best practice guidelines for programming with external calls ([Consensys Recommendations](https://consensys.github.io/smart-contract-best-practices/recommendations/#favor-pull-over-push-for-external-calls), [OpenZeppelin](http://openzeppelin.org/), [Solium Security](https://github.com/duaraghav8/solium-plugin-security), [Mythril](https://github.com/ConsenSys/mythril), [Solcheck](https://github.com/federicobond/solcheck)). This proposal will integrate best practice into the language design. Below are causes for concern with external calls.
+
+#### 1. Contracts are untrustworthy by default
+```javascript
+// bad
+Bank.withdraw(100); // Unclear whether trusted or untrusted
+
+function makeWithdrawal(uint amount) { // Isn't clear that this function is potentially unsafe
+    Bank.withdraw(amount);
+}
+
+// good
+UntrustedBank.withdraw(100); // untrusted external call
+TrustedBank.withdraw(100); // external but trusted bank contract maintained by XYZ Corp
+
+function makeUntrustedWithdrawal(uint amount) {
+    UntrustedBank.withdraw(amount);
+}
+```
+
+#### 2. External calls have arbitrary code execution
+Examples of problems that can arise in Solidity code include:
+##### Race Conditions
+Whether using raw calls (of the form `someAddress.call()`) or contract calls (of the form `ExternalContract.someMethod()`), malicious code might execute. Even if ExternalContract is not malicious, malicious code can be executed by any contracts it calls.
+One particular danger is malicious code may hijack the control flow, leading to race conditions.
+
+##### Re-entrancy
+- `someAddress.send()` and `someAddress.transfer()` in Solidity are considered safe against reentrancy. While these methods still trigger code execution, the called contract is only given a stipend of 2300 gas which is currently only enough to log an event.
+  - Prevents reentrancy but is incompatible with any contract whose fallback function requires 2300 gas or more.
+  - Sometimes the programmer won't want this, but then has to fall back onto the dangerous raw calls.
+
+#### 3. External calls can silently fail
+Solidity offers low-level call methods that work on rawAddress: `address.call()`, `address.callcode()`, `address.delegatecall()`, `address.send()`. These low-level methods never throw an exception so they fail silently.
+
+```javascript
```

Comment:
We should make it clear that these examples are written in solidity, and not a proposal of what Flint code does/should look like.